### ISSUE TITLE
Incremental builds produce blank pages in situations of inconsistent case in permalinks

### ISSUE THREAD

**Issue by gregoryjscott** [jekyll/jekyll#3976]
For example, say you have two pages that define permalinks but they don't define them using the same case. One page defines `permalink: /USA/uppercase` and another page defines `permalink: /usa/lowercase`. The first run of `bundle exec jekyll build` will output the two pages and not complain (/usa/ wins in my test). However, the second run of `bundle exec jekyll build` wipes out the content of _site/usa/uppercase. I threw together a sample repo that shows this commit my commit. See https://github.com/gregoryjscott/inconsistent-permalink-case/commits/master.

I know what you're thinking, why would anyone do that? Good question. Our site, http://fortworthtexas.gov (the site repo is private but the org is https://github.com/cityoffortworth if you're curious), contains content that was migrated to Jekyll from a legacy CMS and the case of the filenames and directories is kind of all over the place. New content is being added with pretty lowercase URLs but when that's mixed in with existing content that's already in directories containing capital letters then this can easily happen. We have well over a hundred pages that now render blank with Jekyll 3 and incremental build enabled.

We can (and will) fix our problem by making all our permalinks consistent, and always explicitly setting the permalink in situations where the source file isn't the desired case. However, the fact that Jekyll has quietly handled this situation fine before incremental builds, and the fact that the side-effect is your content gets wiped out without warning, led me to submit this as a bug.

cc @bechatman

**Comment by alfredxing** [comment]
Hm. Fresh install of the specified bundle, on the first `bundle exec jekyll build`, I get one `usa` folder (lowercase) with both `lowercase.html` and `uppercase.html`. Second run, nothing changes...

**Comment by alfredxing** [comment]
After a bit more investigation, I _think_ I found the issue. It's quite complicated:

- The page's `destination` method returns a destination path with capitalized `USA` (`USA/uppercase/index.html`)
- When writing the actual file though, the system writes to lowercase `usa` (`usa/uppercase/index.html`)
- The incremental `Regenerator` checks, using `File.exist?`, whether each document/page/post's destination path exists. It turns out (at least on my system), `File.exist?` is *case insensitive*.
- So, on the initial `render` (this is all on the second run), incremental regen checks that everything is unmodified and exists, so the pages aren't re-rendered
- Afterward, our handy `Cleaner` builds a list of "obsolete files" by globbing the existing files, then subtracting the "new files" (all of the destination paths accounted for in the current posts/pages/documents list)
- So our obsolete files set includes `usa/uppercase/index.html` (the result of globbing), but our "new files" list contains `USA/uppercase/index.html` (the result of calling `destination` on all the posts/pages)
- Thus, in the end, since the two don't match when subtracting, `usa/uppercase/index.html` is _not_ removed from the obsolete files list, and so it is deleted by the system
- After the clean step comes the write step. This is more or less the red herring in this case. In this step, our Regenerator notices that the file is now gone, and so it forces the uppercase page to write itself (assuming it has been rendered, which it has not). This causes the `usa/uppercase/index.html` file to be re-created with empty contents, creating the appearance that only the contents have been erased for some reason, when in fact the entire file has been deleted and re-created in the process of generating the Jekyll site.

**TL;DR**: The assumptions we make about the case sensitivity of destination paths are inconsistent (or just incorrect) among the many instances where destination paths are handled (in the cleaner, the regenerator, when writing, etc.)

I'm sure we've had a discussion about this before, but I can't recall the result. Are we expecting paths to be handled in a case sensitive manner? If so, is there a way we can enforce that it is written to the FS as such, and a way to check for the existence of the file with case sensitivity in mind?

**Comment by parkr** [comment]
Let's change the behaviour based on the filesystem's case-sensitivity.
